cpp/unreal_plugins/SimulationController/Source/SimulationController/UrdfBotAgent.cpp

```diff
@@ -0,0 +1,179 @@
+                action_space["joint." + joint_pair.first] = std::move(box);
+            }
+        }
+    }
+
+    return action_space;
+}
+
+std::map<std::string, Box> UrdfBotAgent::getObservationSpace() const
+{
+    std::map<std::string, Box> observation_space;
+
+    auto observation_components = Config::get<std::vector<std::string>>("SIMULATION_CONTROLLER.URDFBOT_AGENT.OBSERVATION_COMPONENTS");
+
+    if (Std::contains(observation_components, "link_state")) {
+        for (auto& link_pair : urdf_bot_pawn_->urdf_robot_component_->link_components_) {
+            Box box;
+            box.low_ = std::numeric_limits<float>::lowest();
+            box.high_ = std::numeric_limits<float>::max();
+            box.shape_ = {6};
+            box.datatype_ = DataType::Float32;
+            observation_space["link_state." + link_pair.first] = std::move(box);
+        }
+    }
+
+    std::map<std::string, Box> camera_sensor_observation_space = camera_sensor_->getObservationSpace(observation_components);
+    for (auto& camera_sensor_observation_space_component : camera_sensor_observation_space) {
+        observation_space[camera_sensor_observation_space_component.first] = std::move(camera_sensor_observation_space_component.second);
+    }
+
+    return observation_space;
+}
+
+std::map<std::string, Box> UrdfBotAgent::getStepInfoSpace() const
+{
+    std::map<std::string, Box> step_info_space;
+
+    return step_info_space;
+}
+
+void UrdfBotAgent::applyAction(const std::map<std::string, std::vector<uint8_t>>& actions)
+{
+    auto action_components = Config::get<std::vector<std::string>>("SIMULATION_CONTROLLER.URDFBOT_AGENT.ACTION_COMPONENTS");
+    std::map<std::string, float> joint_actions;
+
+    for (auto& action : actions) {
+        if (Std::contains(action_components, "joint") && action.first.find("joint.") == 0) {
+            std::vector<float> action_data = Std::reinterpret_as<float>(action.second);
+            std::string joint_name = action.first.substr(std::strlen("joint."));
+            joint_actions[joint_name] = action_data.at(0);
+        }
+    }
+
+    urdf_bot_pawn_->urdf_robot_component_->applyAction(joint_actions);
+}
+
+std::map<std::string, std::vector<uint8_t>> UrdfBotAgent::getObservation() const
+{
+    std::map<std::string, std::vector<uint8_t>> observation;
+
+    auto observation_components = Config::get<std::vector<std::string>>("SIMULATION_CONTROLLER.URDFBOT_AGENT.OBSERVATION_COMPONENTS");
+
+    if (Std::contains(observation_components, "link_state")) {
+        for (auto& link_pair : urdf_bot_pawn_->urdf_robot_component_->link_components_) {
+            FVector position = link_pair.second->GetRelativeLocation();
+            FRotator rotation = link_pair.second->GetRelativeRotation();
+            observation["link_state." + link_pair.first] = Std::reinterpret_as<uint8_t>(std::vector<float>{position.X, position.Y, position.Z, rotation.Roll, rotation.Yaw, rotation.Pitch});
```

Comment:
This line is slightly too long. Our current convention is a max of 158 chars per line (unless there is a really good reason) because that is what the GitHub web interface lets you view without line wrapping.

Reply:
fixed